lib/discourse.rb

```diff
@@ -311,25 +311,44 @@ def self.request_refresh!(user_ids: nil)
   def self.git_version
     return $git_version if $git_version
 
-    # load the version stamped by the "build:stamp" task
-    f = Rails.root.to_s + "/config/version"
-    require f if File.exists?("#{f}.rb")
-
-    begin
-      $git_version ||= `git rev-parse HEAD`.strip
-    rescue
-      $git_version = Discourse::VERSION::STRING
-    end
+    git_cmd='git rev-parse HEAD'
+    self.load_version_or_git(git_cmd, Discourse::VERSION::STRING) { $git_version }
   end
 
   def self.git_branch
     return $git_branch if $git_branch
+    git_cmd='git rev-parse --abbrev-ref HEAD'
+    self.load_version_or_git(git_cmd, 'unknown') { $git_branch }
+  end
+
+  def self.full_version
+    return $full_version if $full_version
+    git_cmd='git describe --dirty --match "v[0-9]*"'
+    self.load_version_or_git(git_cmd, 'unknown') { $full_version }
+  end
+
+  def self.load_version_or_git(git_cmd, default_value)
+    version_file  = "#{Rails.root}/config/version.rb"
+    version_value = false
 
-    begin
-      $git_branch ||= `git rev-parse --abbrev-ref HEAD`.strip
-    rescue
-      $git_branch = "unknown"
+    if File.exists?(version_file)
+      require version_file
+      version_value = yield
     end
+
+    # file does not exist or does not define the expected global variable
+    unless version_value
+      begin
+        version_value = `#{git_cmd}`.strip
+      rescue # sollte noch ausspezifiziert werden…
```

Comment:
It would be better to keep the comments in English ;)

Reply:
the worse part is that it is a TODO and i wanted to specify the common exceptions i want to catch instead of the broad sweep handler :P

Just remove the comment or actually limit the exceptions we want to handle there as well?

Reply:
I depends, can we do anything on the exceptions? If not, then it's fine to leave it that way.